Width and Alignment Settings.

uroybd opened the pull request:
More control on positioning and size. Namely:
1. Width Control.
2. Vertical Alignment.
3. Horizontal Alignment.

Hopefully, didn't break anything.

amezin commented:
Positioning still doesn't seem to work correctly for me (dual monitor setup, Wayland):

![Screenshot from 2021-04-22 14-55-08](https://user-images.githubusercontent.com/1273397/115686958-886bc080-a37b-11eb-9be4-11f35a6e7d38.png)

![Screenshot from 2021-04-22 14-59-01](https://user-images.githubusercontent.com/1273397/115686962-8a358400-a37b-11eb-988a-66e683792b0e.png)

amezin commented:
Are you sure you want horizontal/vertical alignment, configurable width and height instead of just "attach to top/bottom/left/right edge"?

uroybd commented:
> Positioning still doesn't seem to work correctly for me (dual monitor setup, Wayland):

Same problem as Yakuake. In Yakuake, if my mouse is in the left monitor and I pull the terminal it comes in the right monitor, but in the left edge. If my mouse is on the right monitor, it comes in the right of the right monitor.

I want to fix it, but I don't have a dual monitor setup in my Home.

> Are you sure you want horizontal/vertical alignment, configurable width, and height instead of just "attach to top/bottom/left/right edge"?

Oh yes. I use the terminal extensively and often with a lot of logs to inspect. It's also easier to compare some logs side by side with some code or snippets of data than top and bottom. This is the optimum terminal size and position for me:

![Screenshot from 2021-04-22 15-14-33](https://user-images.githubusercontent.com/4207230/115689168-a63a2500-a37d-11eb-946c-eb11a5f5d669.png)

amezin commented:
From the screenshot you posted it's not obvious why an option like "Attach to Edge: Right" won't be enough (with "window height" becoming "window size"). Top/bottom/left/right terminal position, with window size configurable only in one direction (height for top/bottom, width for left/right)

amezin commented:
> Same problem as Yakuake. In Yakuake, if my mouse is in the left monitor and I pull the terminal it comes in the right monitor, but in the left edge. If my mouse is on the right monitor, it comes in the right of the right monitor.

No, ddterm always pops up on the monitor with the mouse pointer. On the 1st screenshot, mouse pointer is on the 2nd (vertical) monitor (and the window is there too). On the 2nd screenshot, mouse pointer is on the primary monitor.

Also on the 2nd screenshot you can see that vertical alignment doesn't work on the primary monitor.

uroybd commented:
"Attach to Edge: Right" will be enough for me. But I thought both height and width and alignments in 2d will allow doing more, maybe someone will want the terminal to take the right-bottom corner, only a quarter of one's screen. I wanted to cover a lot of possible use-cases.

uroybd commented:
> > Same problem as Yakuake. In Yakuake, if my mouse is in the left monitor and I pull the terminal it comes in the right monitor, but in the left edge. If my mouse is on the right monitor, it comes in the right of the right monitor.
> 
> No, ddterm always pops up on the monitor with the mouse pointer. On the 1st screenshot, mouse pointer is on the 2nd (vertical) monitor (and the window is there too). On the 2nd screenshot, mouse pointer is on the primary monitor.
> 
> Also on the 2nd screenshot you can see that vertical alignment doesn't work on the primary monitor.

Yeah, it's better than Yakuake!

Does this only happen when there's dual monitors? I think it is solvable only I don't have another monitor in my hand right now.

amezin commented:
> Does this only happen when there's dual monitors?

With single monitor, with "bottom" vertical alignment the window appears at the center:
![Screenshot from 2021-04-22 20-35-41](https://user-images.githubusercontent.com/1273397/115733250-8a00ad00-a3aa-11eb-84af-31fe2be13d34.png)

And "center" is just a few pixels under the top panel

amezin commented on extension.js:
```diff
@@ -314,7 +318,40 @@ function workarea_for_window(win) {
 
 function target_rect_for_workarea(workarea) {
     const target_rect = workarea.copy();
```
Remove this line and simply use `workarea.width`, `workarea.height`, etc. if you recompute all coordinates/dimensions anyway. And then `return new Meta.Rectangle({x, y, width, height});`

uroybd commented:
Done the changes you've asked for.

amezin commented:
See [my previous comment](https://github.com/amezin/gnome-shell-extension-ddterm/pull/32#issuecomment-824900021)

Center and bottom vertical alignment still doesn't work correctly, even with one monitor, on both X11 and Wayland.

Actually, it's because `x = 0` for `left` alignment (and `y = 0` for `top`) isn't correct too. I don't know why they work.. somewhat - even in single monitor case, the top panel makes `y = 0` incorrect. Probably the window manager silently moves the window so that it doesn't overlap with the top panel. It should be `x = workarea.x` (and `y = workarea.y`, correspondingly). And these `x`/`y` offsets should be taken into account for `center` and `bottom`/`right` too.

Also it seems that you skipped multiple `unmaximize()` calls. They all are necessary in one or another configuration. When there is a `.unmaximize(Meta.MaximizeFlags.VERTICAL)`, there should be a similar `.unmaximize(Meta.MaximizeFlags.HORIZONTAL)`. Otherwise, for example, animations break on Wayland when you set the width to something between 80% and 99% (currently I see this effect only on the secondary monitor though).

Also there are a couple of issues with mouse resizing (i. e. dragging the bottom edge):
1) When vertical alignment is set to `Center`, resizing doesn't match cursor movement... And I'm not sure if it can be fixed - resizing is implemented in Gtk and window manager, you can only ask to 'start' it.
2) You'll need to implement mouse resizing for all remaining edges.

uroybd commented:
I've fixed the alignment and sizing issues and added `.unmaximize(Meta.MaximizeFlags.HORIZONTAL)` in some places.

However, It brought a new issue. In preferences, if you drag the height controller (not click, click is fine) it vibrates sometime.

For the last 2 requirements, I would like some guidance.

Also, we can do this incrementally. One at a MR.

amezin commented:
> I've fixed the alignment and sizing issues

"Vertical alignment: Center" is still incorrect: 
![Screenshot from 2021-04-23 13-15-47](https://user-images.githubusercontent.com/1273397/115833625-160be680-a436-11eb-8771-b0a242c332de.png)

uroybd commented:
> > I've fixed the alignment and sizing issues
> 
> "Vertical alignment: Center" is still incorrect:

Weirdly, I can't reproduce the issue. I think I should wait till Gnome 40 and GTK4. It can be due to major version change.

![Screenshot from 2021-04-23 13-34-01](https://user-images.githubusercontent.com/4207230/115836030-cf6bbb80-a438-11eb-9345-829961f34bc6.png)

amezin commented:
Try enabling "Window List" extension (usually comes with Gnome Shell) or any other bottom panel

amezin commented:
> However, It brought a new issue. In preferences, if you drag the height controller (not click, click is fine) it vibrates sometime.

It happens because there are multiple things that update `window-height` (the slider in the preferences dialog, and the extension - for mouse resizing). It is a bug that existed in the code before your pull request, so ignore it. I'm working on a fix.

uroybd commented:
> > I've fixed the alignment and sizing issues
> 
> "Vertical alignment: Center" is still incorrect:

Weirdly, I can't reproduce the issue. I think I should wait till Gnome 40 and GTK4. It can be due to major version change.

![Screenshot from 2021-04-23 13-34-01](https://user-images.githubusercontent.com/4207230/115836030-cf6bbb80-a438-11eb-9345-829961f34bc6.png)

uroybd commented:
> Try enabling "Window List" extension (usually comes with Gnome Shell) or any other bottom panel

Fixed it I hope.

amezin commented:
> It happens because there are multiple things that update window-height (the slider in the preferences dialog, and the extension - for mouse resizing). It is a bug that existed in the code before your pull request, so ignore it. I'm working on a fix.

Fixed in https://github.com/amezin/gnome-shell-extension-ddterm/commit/fef42dcd9ce3336f14cf82374f699452c47a961c
Rebase your branch on top of the current master to get the fix

amezin commented:
When the window is bottom-aligned, there is almost always an one-pixel gap between it and the bottom panel:
![Screenshot from 2021-04-24 00-36-51](https://user-images.githubusercontent.com/1273397/115915786-a4638500-a495-11eb-8cec-6501422b9ffe.png)

amezin commented on extension.js:
```diff
@@ -313,9 +317,50 @@ function workarea_for_window(win) {
-    target_rect.height *= settings.get_double('window-height');
-    return target_rect;
+    const available_width = workarea.width,
+        available_height = workarea.height,
```
What's the point of adding `available_width` and `available_height`? `workarea.width` and `workarea.height` are even shorter by one character.

amezin commented on extension.js:
```diff
@@ -313,9 +317,50 @@ function workarea_for_window(win) {
+        vertical_alignment = settings.get_string('window-vertical-alignment');
+
+    // width = Math.min(full_width * settings.get_double('window-width'), available_width),
+    // height = Math.min(full_height * settings.get_double('window-height'), available_height),
```
Please remove commented-out code.
And, by the way, `settings.get_double('window-height')` (and, I guess, the same is true for `settings.get_double('window-width')`) can't return values greater than 1 - it's defined in the settings schema.

uroybd replied:
Removed.

amezin commented on extension.js:
```diff
@@ -313,9 +317,50 @@ function workarea_for_window(win) {
+    const available_width = workarea.width,
+        available_height = workarea.height,
+        full_width = available_width + workarea.x,
+        full_height =  available_height + workarea.y,
```
What does `full_width` and `full_height` mean? I. e. "full height" of what?

uroybd replied:
Full height of the monitor. x and y are working as kinda offset. so full-height is available height + offset.

amezin replied:
What you're calculating here is not a full height of the monitor. At least if there is a bottom panel. And the same for `full_width` - if the user enables an extension that adds a panel on the right, it won't match the width of the monitor.

amezin commented on extension.js:
```diff
@@ -313,9 +317,50 @@ function workarea_for_window(win) {
+        if (x_diff === 0)
+            x = workarea.x;
+        else
+            x = workarea.x + Math.round(x_diff / 2);
```
I think it should be possible to implement this simpler, without `if` / ternary operator

uroybd replied:
Done.

amezin commented on extension.js:
```diff
@@ -313,9 +317,50 @@ function workarea_for_window(win) {
+        if (y_diff === 0)
+            y = workarea.y;
+        else
+            y = workarea.y + Math.round(y_diff / 2);
```
I think it should be possible to implement this simpler, without `if` / ternary operator

uroybd replied:
Done like before.

amezin commented on extension.js:
```diff
@@ -295,6 +298,7 @@ function track_window(win) {
 
     // https://github.com/amezin/gnome-shell-extension-ddterm/issues/28
     win.connect('shown', update_window_geometry);
+    win.connect('position-changed', update_window_geometry);
```
Please verify if this is still necessary

amezin commented on extension.js:
```diff
@@ -42,7 +42,7 @@ class ExtensionDBusInterface {
         Main.wm.skipNextEffect(current_window.get_compositor_private());
         current_window.unmaximize(Meta.MaximizeFlags.VERTICAL);
-
+        current_window.unmaximize(Meta.MaximizeFlags.HORIZONTAL);
```
This method is called by the application before mouse resizing begins (I guess I should have called it `PrepareForResize` or something like that).
For vertical resizing, `unmaximize()` should be called with `Meta.MaximizeFlags.VERTICAL`. For horizontal resizing - `Meta.MaximizeFlags.HORIZONTAL`. Unmaximizing in both directions at once is not necessary (and maybe it even causes some issues, but I haven't found any so far).
Although there is no horizontal resizing with mouse yet, so for now this line can be simply removed, and then added later, together with horizontal mouse resizing.

uroybd replied:
Removed.

amezin commented on extension.js:
```diff
@@ -330,6 +375,9 @@ function unmaximize_window(win) {
         win.unmaximize(Meta.MaximizeFlags.VERTICAL);
+
+    if (target_rect.width < workarea.width)
+        win.unmaximize(Meta.MaximizeFlags.HORIZONTAL);
```
`unmaximize_window` is only called when the window is maximized vertically: `win.connect('notify::maximized-vertically', unmaximize_window);`
`win.unmaximize(Meta.MaximizeFlags.HORIZONTAL);` should be in another function, called in response to `notify::maximized-horizontally`

uroybd replied:
renamed `unmaximize_window` to `unmaximize_window_vertically` and added `unmaximize_window_horizontally` and connected and disconnected them in appropriate places.

uroybd commented:
@amezin have you changed something in the `Makefile`?

`WITH_GTK4=no` is not working.

uroybd reviewed (commented):
Mostly Done. Couldn't test, build issues.

amezin commented:
Please rebase properly. There should be no [merge commits](https://github.com/amezin/gnome-shell-extension-ddterm/pull/32/commits/d615caeefc0866fc96325db603309fc212d37d5c).

amezin commented:
> @amezin have you changed something in the Makefile?
> 
> WITH_GTK4=no is not working.

No, I did not change anything in Makefile. What do you mean by "not working"? What error are you seeing?

uroybd commented:
> > @amezin have you changed something in the Makefile?
> > WITH_GTK4=no is not working.
> 
> No, I did not change anything in Makefile. What do you mean by "not working"? What error are you seeing?

Here:s the error:
```
/bin/bash: line 1: gtk4-builder-tool: command not found
make: *** [Makefile:59: gtk-builder-validate/prefs-gtk4.ui] Error 127
```

uroybd commented:
> Please rebase properly. There should be no [merge commits](https://github.com/amezin/gnome-shell-extension-ddterm/pull/32/commits/d615caeefc0866fc96325db603309fc212d37d5c).

I think it's fine now.

amezin commented:
> Here:s the error:
> 
> ```
> /bin/bash: line 1: gtk4-builder-tool: command not found
> make: *** [Makefile:59: gtk-builder-validate/prefs-gtk4.ui] Error 127
> ```

It means that you have created `prefs-gtk4.ui` somehow. If you remove it, the error will disappear.

After https://github.com/amezin/gnome-shell-extension-ddterm/commit/0c6e7afd438ec23ea89ef4b9f286e68852874713 it won't validate `prefs-gtk4.ui` when `WITH_GTK4=no` even if the file exists.

amezin commented on extension.js:
```diff
@@ -312,12 +316,41 @@ function workarea_for_window(win) {
+        break;
+    case 'center':
+        var x_diff = workarea.width  - width;
+        x = workarea.x + (x_diff === 0) ? 0 : Math.round(x_diff / 2);
```
It doesn't look any simpler than the previous version. Why do you need the `if`/ternary operator? Isn't `0 / 2` equal to `0`?

uroybd replied:
Done.

amezin commented on extension.js:
```diff
@@ -42,7 +42,6 @@ class ExtensionDBusInterface {
 
         Main.wm.skipNextEffect(current_window.get_compositor_private());
         current_window.unmaximize(Meta.MaximizeFlags.VERTICAL);
-
```
Please remove whitespace-only changes (i. e. there should be no blocks where only an empty line is added or removed)

amezin replied:
Sorry, maybe I wasn't clear enough. I didn't ask to reformat the whole file. I asked to get rid of changes like this line: where you remove (or add) an empty line in the middle of the code that is otherwise unchanged.

uroybd replied:
You were perfectly clear.

What's happening is:
* My editor uses prettier as default. So I disable formatting altogether for this project so that existing formatting doesn't break.
* I saw that you're using eslint, so before push, I run `eslint --fix` so that all the files get formatted in your way. However some lines are getting like the one you review, I don't know why. So I jut reverted them to original styles.

amezin commented on glade/prefs.ui:
```diff
@@ -593,6 +599,64 @@
                 <property name="top-attach">1</property>
               </packing>
+            </child>
+                        <child>
```
What's going on with formatting here? Is Glade doing this?

uroybd replied:
Done.

amezin commented:
Please rebase your pull request again. I've simplified the code related to mouse resizing a lot, you'll need that later.

Rebase will be easier if you:
1) get rid of whitespace-only changes
2) squash all your commits into one

amezin commented on extension.js:
```diff
@@ -331,6 +364,20 @@ function unmaximize_window(win) {
+    if (!win || win !== current_window)
+        return;
+
+    if (!win.maximized_vertically)
```
I think it should be `maximized_horizontally` here

uroybd replied:
What a stupid mistake! >.<
Fixed.

uroybd replied:
Yes. Very Stupid mistake. Fixed.

amezin commented on extension.js:
```diff
@@ -22,15 +22,23 @@ let grab_window = null;
+        this.dbus = Gio.DBusExportedObject.wrapJSObject(
+            ByteArray.toString(xml),
+            this
+        );
```
Why do you reformat the code in a way that obviously doesn't match the rest of the file? And why do you reformat the code that you didn't need to modify at all?

amezin commented on extension.js:
```diff
@@ -22,15 +22,23 @@ let grab_window = null;
+const SUBPROCESS_ARGV = [
+    Me.dir.get_child('com.github.amezin.ddterm').get_path(),
+    '--undecorated',
+];
```
Please do not reformat the code unnecessarily. It only creates additional conflicts when merging/rebasing.

amezin reviewed (changes requested):
https://github.com/amezin/gnome-shell-extension-ddterm/pull/32#discussion_r619716034

What about https://github.com/amezin/gnome-shell-extension-ddterm/pull/32#issuecomment-825848639 ?

Please get rid of whitespace-only/reformatting changes, squash your commits into one commit, and rebase on the current `master` branch.

uroybd reviewed (commented):
I think, I've fixed the formatting issue.

uroybd commented:
Hi,

No more whitespace or formatting diffs.

uroybd reviewed (commented):
All done as you requested.

amezin commented:
@uroybd 

> All done as you requested.

No, it's not

https://github.com/amezin/gnome-shell-extension-ddterm/pull/32#pullrequestreview-644155609

> squash your commits into one commit, and rebase on the current master branch.

There are currently 21 commits, and many of them are just reverting parts of other commits from the same pull request. https://about.gitlab.com/blog/2018/06/07/keeping-git-commit-history-clean/ Please, squash all your commits into one commit.

uroybd commented:
Squashed and Rebased.

amezin commented on toggle.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
This file was [added](https://github.com/amezin/gnome-shell-extension-ddterm/commit/e0456a8d8968361fa69065d59a9b07d0e1e71c65#diff-702eeb57673f7a7a4aa1547229d25069e01a60bc975daf536742a423bffd14e9), and then [removed](https://github.com/amezin/gnome-shell-extension-ddterm/commit/834f7aece230944175a65fc22693ac692a2c6551#diff-702eeb57673f7a7a4aa1547229d25069e01a60bc975daf536742a423bffd14e9) on the `master` branch. It should not be in your pull request.

uroybd replied:
Removed.

amezin commented on extension.js:
```diff
@@ -321,12 +325,39 @@ function workarea_for_window(win) {
-    target_rect.height *= settings.get_double('window-height');
-    return target_rect;
+    const full_width = workarea.width + workarea.x,
+        full_height = workarea.height + workarea.y,
```
Again, `workarea.height + workarea.y` isn't a "full height" of anything. Sometimes it matches the height of the monitor, but only if there is no bottom panel.

Please rename this variable, or remove it - it's only used once.

Exactly the same for `full_width`

uroybd replied:
Done.

I'm not much familiar with gjs so a stupid question: How can I actually get the full height and width of the monitor?

amezin replied:
https://gjs-docs.gnome.org/meta7~7_api/meta.display#method-get_monitor_geometry

However, you shouldn't need it

amezin reviewed (changes requested):
https://github.com/amezin/gnome-shell-extension-ddterm/pull/32#issuecomment-825848639 this issue is still present

The same happens when horizontal alignment is set to "Right" (one pixel gap between the right edge of the window and the right edge of the screen). The gap appears and disappears when window width/height changes.

uroybd commented:
> [#32 (comment)](https://github.com/amezin/gnome-shell-extension-ddterm/pull/32#issuecomment-825848639) this issue is still present
> 
> The same happens when horizontal alignment is set to "Right" (one pixel gap between the right edge of the window and the right edge of the screen). The gap appears and disappears when window width/height changes.

FIxed it. Rounding was required for `height` and `width`.

amezin reviewed (changes requested):
Window positioning seems to work fine now.

Now it's time for resizing with mouse. Currently there are 2 kinds of issues:
1) you can always drag the bottom edge of the window, even when the window is bottom-aligned
2) you can't drag any other edge

I'm still not sure if separate horizontal/vertical alignment and resizing in both directions are OK.

One example why I like simpler "attach to edge" + generic "window size" parameter (height for top/bottom alignment, width for left/right) more:
- You have terminal attached to the top edge, with window size (height in this case) set to 60%.
- You change terminal attachment to "right edge"
- Now you have terminal attached to the right edge, with width automagically set to 60% ("window size" parameter changed its meaning - from "height" to "width" - but not its value).

Yes, you can achieve the same effect with separate width, height, alignment, by manually swapping with and height values when alignment is changed, but it will likely be confusing.

Also, I'm not sure if resizing with mouse could be implemented correctly for a centered window (and it's not "how hard/complex it will be", it's "not sure if doable at all")

Also, when I see a borderless terminal window in the corner of the screen, it just looks like something is broken :)

uroybd commented on prefs.js:
```diff
@@ -161,6 +165,10 @@ function createPrefsWidgetClass(resource_path, util) {
+                this.settings_bind('window-width', this.window_width_adjustment, 'value');
+                this.set_scale_value_format_percent(this.window_width_scale);
+                this.settings_bind('window-vertical-alignment', this.window_vertical_alignment, 'active-id');
+                this.settings_bind('window-horizontal-alignment', this.window_horizontal_alignment, 'active-id');
```
Done.

uroybd commented on prefs.js:
```diff
@@ -161,6 +165,10 @@ function createPrefsWidgetClass(resource_path, util) {
 
+                this.settings_bind('window-width', this.window_width_adjustment, 'value');
+                this.set_scale_value_format_percent(this.window_width_scale);
+                this.settings_bind('window-vertical-alignment', this.window_vertical_alignment, 'active-id');
```
Done.

uroybd commented:
'Attach to Edge' is a very good solution. However, as I said earlier, it gives you a little less control.

> Also, when I see a borderless terminal window in the corner of the screen, it just looks like something is broken :)

It's kind of a personal preference, right? Like, in my home I use a laptop with a small screen so I like it in 100% height, 40% width on right. However, in my office I have 3 monitors, one is pretty big. I really don't need a very big terminal. So keep it in the bottom-right corner, taking 50% height.

uroybd commented:
'Attach to Edge' is a very good solution. However, as I said earlier, it gives you a little less control.

> Also, when I see a borderless terminal window in the corner of the screen, it just looks like something is broken :)

It's kind of a personal preference, right? Like, in my home I use a laptop with a small screen so I like it in 100% height, 40% width on right. However, in my office I have 3 monitors, one is pretty big. I really don't need a very big terminal. So keep it in the bottom-right corner, taking 50% height.

I might need some help regarding `resizing_box`. How can I implement it for other sides?

amezin commented:
I prefer "Attach to edge" (https://github.com/amezin/gnome-shell-extension-ddterm/issues/35), because it's simpler (mainly easier to test).

This looks too complex and takes too much time.

Won't accept, sorry

namefailed commented:
What would need to be changed to bring this back to life?

amezin commented:
This never worked fully correctly in the first place, so "bringing it back" won't help.

To make it possible, you need to fix things like this https://gitlab.gnome.org/GNOME/mutter/-/issues/1627 in GNOME's window manager, and, probably, introduce an additional API to disable all "smart" logic (like auto-maximization, window placement) for select windows.
